covdefaults.py

```diff
@@ -84,6 +98,81 @@ def configure(self, config: CoverageConfig) -> None:
         if not config.get_option('report:fail_under'):
             config.set_option('report:fail_under', 100)
 
+    def fix_omit(self, config: CoverageConfig) -> None:
+        # subtract omit settings if requested
+        omit = set(config.get_option('run:omit') or [])
+        omit.add('*/setup.py')  # always ignore setup.py
+
+        # ignore common virtual environment folders, unless it's within source
+        source = config.get_option('run:source')
+        sep, suf = os.sep, ['.py', '.pyc', '.pyo']
+        for folder, is_prefix in (
+                ('.tox', False),
+                ('.nox', False),
+                ('venv', True),
+                ('.venv', True),
+        ):
+            matcher = re.compile(f'.*{re.escape(f"{sep}{folder}{sep}")}.*')
+            for src in source:
+                if matcher.match(src) is not None:
+                    # now we can't cover this with one omit,
+                    # climb the tree and exclude everything else
+                    parts = os.path.relpath(src, os.getcwd()).split(sep)
+                    for at in range(1, len(parts)):
+                        level = sep.join(parts[:at])
+                        child = parts[at]
+                        for entry in os.listdir(level):
+                            if entry == child:  # dont' exclude children
+                                continue
+                            pattern = f'*/{level}/{entry}'
+                            if os.path.isdir(os.path.join(level, entry)):
+                                pattern += '/*'
+                            else:
+                                if not any(entry.endswith(e) for e in suf):
+                                    continue
+                            omit.add(pattern)
+                    break
+            else:  # blank omit any such folders
+                omit.add(f'*/{folder}{"*" if is_prefix else ""}/*')
+
+        omit.difference_update(self._subtract_omit)
+        config.set_option('run:omit', sorted(omit))
+
+    def _set_source(self, config: CoverageConfig) -> None:
+        if not isinstance(config, CoverageConfig):  # pragma: no cover
+            # https://github.com/nedbat/coveragepy/issues/967
+            config = config.config
```

Comment:
we should not do this, this adds too much fraglity -- the api is only `set_option` and `get_option` -- do we not have access to `set_option('paths:source')`?

Reply:
it does not work, raised https://github.com/nedbat/coveragepy/issues/967 for upstream, but that's unlikely to be handled for ages...

Reply:
send a patch?  I can't use private implementation details in something that's going to suddenly break everyone's CI

Reply:
https://github.com/nedbat/coveragepy/pull/969 🤷‍♂ not sure when that will land